### ISSUE TITLE
Preprocess Imputer always returning output using mean strategy

### ISSUE THREAD

**Issue by manish7294** [mlpack/mlpack#1183]
Preprocess Imputer currently imputing the missing values using only mean strategy even-though user provides different strategies. 
Run Commands Used:

> mlpack_preprocess_imputer -i preprocess_imputer_test.csv -o result.csv -m nan -d 0 -s listwise_deletion

> mlpack_preprocess_imputer -i preprocess_imputer_test.csv -o result.csv -m nan -d 0 -s median 

_Most probably the instance of imputer is not updating the change in strategy on re-initiating as I have confirmed this by changing default strategy value._

**Comment by rcurtin** [comment]
I can confirm that this is an issue by looking at the code.  Apparently this slipped by on review.  This is out of preprocess_imputer_main.cpp:

```
    // Initialize imputer class
    Imputer<double, MapperType, MeanImputation<double>> imputer(info);
    if (strategy == "mean")
    {
      Imputer<double, MapperType, MeanImputation<double>> imputer(info);
    }
    else if (strategy == "median")
    {
      Imputer<double, MapperType, MedianImputation<double>> imputer(info);
    }
    else if (strategy == "listwise_deletion")
    {
      Imputer<double, MapperType, ListwiseDeletion<double>> imputer(info);
    }
    else if (strategy == "custom")
    {
      CustomImputation<double> strat(customValue);
      Imputer<double, MapperType, CustomImputation<double>> imputer(
          info, strat);
    }
    else
    {
      Log::Fatal << "'" <<  strategy << "' imputation strategy does not exist"
          << endl;
    }
```

but this is very flawed because `imputer` is being declared at two different scopes.  The outer scope `imputer` is the one that ends up being used for imputation, and that is always using the mean strategy.  The block of code will need to be rewritten to properly handle `imputer` having different types (possibly by enclosing the actual call to `Impute()` in those inner if statements).

Would you like to fix this as part of the preprocessing binding tests you have in PR #1174?

**Comment by manish7294** [comment]
That's exactly what's going on. Sure, I will fix this within #1174.
